src/librustc/middle/def.rs

```diff
@@ -125,25 +125,29 @@ impl Def {
         }
     }
 
-    pub fn def_id(&self) -> DefId {
+    pub fn def_id_opt(&self) -> Option<DefId> {
         match *self {
             Def::Fn(id) | Def::Mod(id) | Def::ForeignMod(id) | Def::Static(id, _) |
             Def::Variant(_, id) | Def::Enum(id) | Def::TyAlias(id) | Def::AssociatedTy(_, id) |
             Def::TyParam(_, _, id, _) | Def::Struct(id) | Def::Trait(id) |
             Def::Method(id) | Def::Const(id) | Def::AssociatedConst(id) |
             Def::Local(id, _) | Def::Upvar(id, _, _, _) => {
-                id
+                Some(id)
             }
 
             Def::Label(..)  |
             Def::PrimTy(..) |
             Def::SelfTy(..) |
             Def::Err => {
-                panic!("attempted .def_id() on invalid def: {:?}", self)
+                None
             }
         }
     }
 
+    pub fn def_id(&self) -> DefId {
+        self.def_id_opt().expect(&format!("attempted .def_id() on invalid def: {:?}", self))
```

Comment:
Nit: put these on two separate lines (lining up the `.`) -- seems a bit long for one line to me